parquet-column/src/main/java/org/apache/parquet/column/statistics/Statistics.java

```diff
@@ -150,14 +190,16 @@ public int hashCode() {
   public void mergeStatistics(Statistics stats) {
     if (stats.isEmpty()) return;
 
-    if (this.getClass() == stats.getClass()) {
+    // Merge stats only if they have the same type and comparator (the sorting order
+    // is the same)
+    if (this.getClass() == stats.getClass() && Objects.equals(comparator(), stats.comparator())) {
```

Comment:
I think this check can be stronger. Rather than checking the comparator's equality, why not check the underlying type? This check would allow merging stats for binary and binary/UTF8 types, but I don't think that's right.

Reply:
In your example the comparators would not be the same. That's the point here. If the comparators are the same then it does not matter if the two types are different or not as we are calculating min/max in the same way. Am I missing something?

Reply:
Binary and strings use the same comparator so it would allow merging. I don't think it should allow merging binary and string stats because something has gone wrong if we are trying to.

Reply:
That was the original behaviour and I didn't change that. I you think we shall improve this logic here then I have to save the actual Type object to the Statistics. (Currently, type is only used to get the comparator.)